### PR TITLE
[v2.0.0] Build process: linting, compiling and bundling

### PR DESCRIPTION
**Pull request by phamann**
**⚠️ Note this is a PR into the v.2.0.0 branch and not master**

### TL;DR
Adds the new webpack based build process for v2. Compiles source TypeScript files in `src` directory to JavaScript bundles in `dist` via babel. Has separate configuration for development and production workflows including a development server with hot reload.

### What?
Before we begin work on v2 we need a stable development workflow and build process. All source in v2 will be written in [TypeScript](https://www.typescriptlang.org/) which is a compiled language so we need a compile process, we also want to transpile the output to add feature support for our [supported browsers](). 

This PR adds [webpack](https://webpack.js.org/) and its required dev dependencies, separating the configuration into three files: 
- `webpack.common.js` for shared
- `webpack.dev.js` for development
- `webpack.prod.js` for production.

We use the [`babel-loader`](https://github.com/babel/babel-loader) to handle TypeScript compilation and feature transpilation and polyfilling via [core-js](https://github.com/zloirock/core-js). Configuration for this can be found in the `.babelrc` file. As per the v1 setup we choose not to polyfill certain large APIs as we will gate aginst them in the init process.

As we are moving all logic to run inside a WebWoker in v2, we use the [`workerize-loader`](https://github.com/developit/workerize-loader) to emit the worker.ts module as a separate bundle chunk and automatically load and reference the file in the main bundle.

Compiler verification, linting and pretty printing is handled via [ESLint](https://eslint.org/) with its TypeScript and Prettier plugins. Configuration for this can be found in the `.eslintrc` file. We use the webpack [`eslint-loader`](https://github.com/webpack-contrib/eslint-loader) to run this process on every build and emit warnings and errors to the console, however we have also exposed via an npm script `lint` for convenience. 

For development, we use [`webpack-dev-server`](https://github.com/webpack/webpack-dev-server) to automatically generate, serve and open an index.html file with the latest build, this will hot reload with changes for a nicer development experience. The server can be started via an npm script `start:dev`.

Lastly, webpack handles the transclusion of a license banner at the top of all emitted bundles in production.

### Note
I have included scaffolding of the project entry points `index.ts` and `woker.ts` in this PR to test and validate the workflow is working as expected, these currently contain debug log statements which will be removed in later PR.

```
dist
├── 3b64f31ed45199bf71e3.worker.js
├── index.html
└── main.f9899c1d9fa77d7bfc6c.js
src
├── index.html
├── index.ts
└── worker.ts
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,10 +4,36 @@
   "main": "dist/lib.js",
   "license": "MIT",
   "scripts": {
-    "test": ""
+    "test": "",
+    "start:dev": "webpack-dev-server --progress --open --config webpack.dev.js",
```
Comment (clight-fastly) [inline package.json]: using `start` would be nice as it's typically expected

File: README.md [hunk 2]
```diff
@@ -13,33 +13,27 @@ Real user monitoring
...
 - Node.js >= 6 (```brew install node```)
-- Yarn (```brew install yarn```)
 
 ### Install
 ```sh
 git clone git@github.com:fastly/insights.js.git
 cd insights.js
-yarn
-yarn run build
+npm install
```
Comment (clight-fastly) [inline README.md]: Why using `npm` instead of `yarn` ?

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+  "compilerOptions": {
```
Comment (clight-fastly) [inline tsconfig.json]: Why no `strict` option here?

File: .eslintrc [hunk 4]
```diff
@@ -1,44 +1,46 @@
 {
-    "extends": ["prettier"],
```
Comment (clight-fastly) [inline .eslintrc]: Is there a reason for using eslint over tslint?

File: webpack.prod.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+   const date = new Date(now).toISOString();
+   return license.replace('[year]', year).replace('[date]', date);
+}
+
+module.exports = merge(common, {
+   mode: 'production',
+   plugins: [
+      new webpack.BannerPlugin({
+         banner: getLicenseWithDate()
+      })
```
Comment (clight-fastly) [inline webpack.prod.js]: Would suggest adding the bundle analyzer here for reviewing before merge:
https://github.com/webpack-contrib/webpack-bundle-analyzer

File: .eslintrc [hunk 6]
```diff
@@ -1,44 +1,46 @@
...
+        sourceType: "module"
+    },
+    env: {
+		es6: true,
+		browser: true
 	},
-	"parserOptions": {
-		"sourceType": "module",
-        "ecmaVersion": 2017
+    rules: {
```
Comment (oso-mate) [inline .eslintrc]: We use 2 spaces for indentation in other JS apps, no strong opinions though.
We tried to have shareable eslint/prettier configurations, but they ended up in the [monorepo](https://github.com/fastly/ux-platform), if any interest in sharing rules across projects, we should consider put these back to single repos.

Comment (clight-fastly) [reply .eslintrc]: +1 for `"tabWidth": 2`

File: README.md (hunk 2 above)
Comment (phamann) [reply README.md]: Personally, i've found that since version 6, `npm` has been very fast comparatively. And it's one less dependency for consumers to worry about. It also follows convention with other repos within Fastly.

File: .eslintrc (hunk 4 above)
Comment (phamann) [reply .eslintrc]: As `eslint` is now officially adopted by the TypeScript team and compatibility has improved dramatically. https://eslint.org/blog/2019/01/future-typescript-eslint

File: LICENSE [hunk 7]
```diff
@@ -1,8 +1,9 @@
 Fastly Insights.js
-Build generated: <%= moment().format('YYYY-MM-DD') %>
+Build: [hash]
+Generated: [date]
 https://github.com/fastly/insights.js
 
-Copyright (c) <%= moment().format('YYYY') %>, Fastly, Inc. All rights reserved.
+Copyright (c) [year], Fastly, Inc. All rights reserved.
```
Comment (oso-mate) [inline LICENSE]: Cool, didn't know this bracket syntax, much cleaner.

File: webpack.prod.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
+const merge = require('webpack-merge');
+const common = require('./webpack.common.js');
+
+function getLicenseWithDate() {
+   const license = fs.readFileSync(path.resolve(__dirname, 'LICENSE'), { encoding: 'utf8'});
+   const now = Date.now();
+   const year = new Date(now).getFullYear();
+   const date = new Date(now).toISOString();
+   return license.replace('[year]', year).replace('[date]', date);
```
Comment (oso-mate) [inline webpack.prod.js]: Ah, I see how this works now. 👍

File: LICENSE (hunk 7 above)
Comment (phamann) [reply LICENSE]: Some of them are provided by webpack banner plugin for free, such as `[hash]`. So I just replicated that syntax to add the year and date.

File: .eslintrc (hunk 4 above)
Comment (psbanka) [reply .eslintrc]: Nice! Did not know that. Will have to switch back to eslint.

File: README.md (hunk 2 above)
Comment (psbanka) [reply README.md]: agreed
